### DISCUSSION TITLE
MFA Error - Must be executed with ADFS Administration rights granted part 2

### DISCUSSION THREAD

**Discussion by Maurice-AxiansNL** [neos-sdi/adfsmfa#220] [Q&A] [answered]
Hi guys, I tried all suggestions about this issue but can't solve it.

I followed these steps:

Alle steps are on a fresh ADFS 2019 server:

1. Before the setup I added a Active directory security group (ADFS Delegation) to the local administrator group on the ADFS server and added the group in the "ADFS Federation Service properties" in the "Enable delegation for service for administation" part

![image](https://user-images.githubusercontent.com/70941951/154471783-88211adb-c649-4016-b15c-e43857d63175.png)

2. After added the groups I rebooted the server
3. After the reboot I started adfsmfa.3.1.2202.1 and followed the steps
4. After the installation I checked if the system.db already exists (not true)
5. Open Powershell as administrator -> Register-MFASystem

After registering I've got a strange warning / error:

![image](https://user-images.githubusercontent.com/70941951/154471049-8e184dcd-5ada-44de-b5b9-188d14f5ff39.png)

6. Check if the system.db exists -> True
7. Tried some other powershell cmdlets but all return with the same strange Unhandled exception warning / error.
8. Reboot server
9. Start MFA MMC -> Returns with error "Must be executed with ADFS administration rights granted"

I'm lost, can someone point me in de right direction? As you can see I followed the instructions from the installation part and the suggestions from another thread but still no solution.

**Comment by redhook62** [comment]
Hi

The points indicated from 1 to 5 are correct, but why reboot the server?
You have a big problem when trust relationships are requested "AccessViolationException - Memory may be corrupted!"

At any rate. you can delete the System.db, config.db, and blob.db files on each server.
Then run **restart-service mfanotifhub.**
You should see, if all the prerequisites are respected (such as the opening of the firewall rules) these files, on all the servers.
Config.db - MFA configuration cache
System.db - ACL cache for service access (MMC, PS)
Blob.db - MDS file from fido alliance (downloaded).

The account used must have the privileges to browse the different ADDS trust relationships
You can specify this account in the Security tab

Your users using MMC or Powershell must be local/domain administrators to manage all features

See

https://github.com/neos-sdi/adfsmfa/wiki/01-Installation#prerequisites
https://github.com/neos-sdi/adfsmfa/wiki/01-Installation#configure-windows-firewall-rules


regards

**Reply by Maurice-AxiansNL** [reply]
Thanks, I'll give it a try next week on a fresh installed ADFS host.

**Reply by Maurice-AxiansNL** [reply]
Very strange, I installed a new 2019 vm with ADFS configured. Meet the requirements, installed the MFA software and after the register-mfasystem command I got the same Unhandled exeption warning / error as the screenshot above. Any idea's?

blob / config / system.db exists. I can open the MFA mmc but i'm sure I can't anymore after rebooting the VM.

Restarting the mfanotifhub does resulting in the "Must be executed with ADFS Administrative rights granted for current user"

Removing the db files and restart the server indeed creates the db files but still the same Administrative rights error.

**Reply by Maurice-AxiansNL** [reply]
Oke it's clear now. The "Must be executed with ADFS administrative rights" warning is because the ADFS service is crashing after running the setup of the MFA software. Permissions can't be checked if the service isn't running. I can't fix the issue to start it again, if I remove the MFA software the ADFS service is running perfectly.

Eventlog:

**Event ID 7034:** The Active Directory Federation Services service terminated unexpectedly.  It has done this 6 time(s).

&

**Event ID 1000:**

Faulting application name: Microsoft.IdentityServer.ServiceHost.exe, version: 10.0.17763.1131, time stamp: 0x89ebc059
Faulting module name: VCRUNTIME140_CLR0400.dll, version: 14.10.25028.0, time stamp: 0x5bac17be
Exception code: 0xc0000005
Fault offset: 0x000000000000c690
Faulting process id: 0x1638
Faulting application start time: 0x01d82d821c3bc941
Faulting application path: C:\Windows\ADFS\Microsoft.IdentityServer.ServiceHost.exe
Faulting module path: C:\Windows\SYSTEM32\VCRUNTIME140_CLR0400.dll
Report Id: 22108bd3-917c-4424-b847-63e5f28661eb
Faulting package full name: 
Faulting package-relative application ID: 

**There is an internal error from the ADFS service and not from the MFA.**
**Make sure that as indicated in the documentation that the Framework 4.7.2 is installed**

&

**Event ID 1026:**

Application: Microsoft.IdentityServer.ServiceHost.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
   at _**System.Runtime.InteropServices.Marshal.CopyToManaged(IntPtr, System.Object, Int32, Int32)**_
   at _System.DirectoryServices.ActiveDirectory.ForestTrustRelationshipInformation.GetForestTrustInfoHelper()_
   at _System.DirectoryServices.ActiveDirectory.ForestTrustRelationshipInformation.get_TopLevelNames()_
   at **Neos.IdentityServer.MultiFactor.Data.ADDSUtils.LoadForests(System.String, System.String, System.String, Boolean, Boolean)**
   at Neos.IdentityServer.MultiFactor.AuthenticationProvider.OnAuthenticationPipelineLoad(Microsoft.IdentityServer.Web.Authentication.External.IAuthenticationMethodConfigData)
   at Microsoft.IdentityServer.Web.Authentication.External.ExternalAuthenticationHandlerBase.OnAuthenticationPipelineLoad(Microsoft.IdentityServer.Web.Authentication.External.IAuthenticationMethodConfigData)
   at Microsoft.IdentityServer.Web.Authentication.AuthenticationMethodLoader.LoadAuthenticationHandler(Microsoft.IdentityServer.PolicyModel.Configuration.AuthenticationMethodDescriptor, System.String)
   at Microsoft.IdentityServer.Web.Authentication.AuthenticationMethodLoader.LoadAuthenticationHandlers(System.Collections.Generic.IEnumerable`1<Microsoft.IdentityServer.PolicyModel.Configuration.AuthenticationMethodDescriptor>, System.String)
   at Microsoft.IdentityServer.Web.PassiveProtocolListener.LoadAuthenticationHandlers()
   at Microsoft.IdentityServer.Web.PassiveProtocolListener.InitializePipeline()
   at Microsoft.IdentityServer.Web.PassiveProtocolListener.Start()
   at Microsoft.IdentityServer.ServiceHost.STSService.StartListener(System.Type, Int32, Int32, Boolean, Boolean, Boolean, System.String)
   at Microsoft.IdentityServer.ServiceHost.STSService.OnStartInternal(Boolean)
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

**The stack indicates that it is when you retrieve your trust relationships between ADDS forests that the error occurs.**
**On the other hand "CopyToManaged" is an operating system error.**
**And all this happens when starting the MFA.**

**The problem must come from the account you are using.**
**If it does not have the required rights on the remote forests, the recovery of trust relationships fails**
**as stated in the documentation, either everything runs on the ADFS Service account (not recommended), or you must** **specify a data access account (ADDS or SQL depending on your configuration), for ADDS the account of ADFS service or** the data access account must have read/write rights to all users in ALL affected forests, and of course have browse rights to** **those forests. if you don't agree with this choose the SQL Server configuration (anyway it's the best solution for ADFS for** **SAML)**

**Then, whether it is the interactive user (whatever his rights), the service account, or an operator. they must be a member** **of the local administrators group or the ADFS management group defined in the ADFS console.**

**Check and apply all prerequisites.**
**Choose between ADDS mode (rights on all forests) or SQL Server (ADFS service and MFA Service).**

**The only thing I can confirm is that the error occurs in a Microsoft module that is not directly called by MFA, really look at** **your trust relationships, disable MFA and try to connect with a user from another ADDS forest**

**To try to get more information**
**put this in Microsoft.IdentityServer.Servicehost.exe.config directory C:\windows\ADFS**
```xml
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
    ...
   <runtime>
     <legacyCorruptedStateExceptionsPolicy enabled="true"/>
   </runtime>
    ...
</config>
```

**regards**

**redhook**

**Reply by Maurice-AxiansNL** [reply]
Weird, the service account ADFS service is running is domain admin and local admin. The account running the MFA setup is either domain admin and local admin en part of the Security group that has access to ADFS delegation. 

I use local DB for ADFS and don't use SQL for MFA software.

**Reply by redhook62** [reply]
Confirm that when the mfa is disabled (in the ADFS console), you can log in with:
- A domain admin account of the domain
- A domain user account
- A Domain Admin account from a remote forest (Trust RelationShip)
- A User Account from a remote forest

The problem is related to your trust relationships between your differents forests/domains

Application: Microsoft.IdentityServer.ServiceHost.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
at **System.Runtime.InteropServices**.Marshal.CopyToManaged(IntPtr, System.Object, Int32, Int32)
at **System.DirectoryServices.ActiveDirectory**.ForestTrustRelationshipInformation.GetForestTrustInfoHelper()  **<-- Here**
at **System.DirectoryServices.ActiveDirectory**.ForestTrustRelationshipInformation.get_TopLevelNames()
at _Neos.IdentityServer.MultiFactor.Data.ADDSUtils.LoadForests(System.String,_ System.String, System.String, Boolean, Boolean)
at Neos.IdentityServer.MultiFactor.AuthenticationProvider.OnAuthenticationPipelineLoad(Microsoft.IdentityServer.Web.Authentication.External.IAuthenticationMethodConfigData)

Check and validate all your trust relationships at the level of your ADDS infrastructure

have you integrated the "Runtime" tag indicated above into the configuration file. ?

regards

**Reply by Maurice-AxiansNL** [reply]
Hi,

I restored a snapshot of the machine and tried the runtime parameter. Now both services will start. I can start the MFA console now :)

**Reply by redhook62** [reply]
Have you some traces in eventlog ?
This runtime parameter should not be set, it' only for debugging.

**Reply by Maurice-AxiansNL** [reply]
Event 5100 ADFS MFA DataServices: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

**Reply by redhook62** [reply]
putting the runtime tag in the configuration file simply prevents the ADFS service from crashing, however the Framework does generate an AccesViolationException error, so MFA cannot detect TopLevelNames and ADDS trust relationships.
This is an internal Framework error. 
The account executing this code (ADFS account), does not have the necessary rights to access and browse remote forests. and Microsoft unmanaged code crashes miserably.

The idea is to remove the tag.
Can you in the console, in the security tab, indicate an ADDS access account (local forest and remote forests) with read/write/traverse rights on all forests.
On the other hand, try with this account (Super User) to connect to all forests with a tool like LDAPAdmin

regards

**Comment by redhook62** [comment, marked as answer]
Hi @mrblwd-058 

Yes, it's a Microsoft update problem.

If you patched your _domain controllers_ with **KB5011495**, the other servers accessing .domain controllers and requesting trust relationships with .Net framework, fail miserably and crash as indicated in your post.

The Framework of the machines concerned (your ADFS servers or other) must apply a patch that is not published by Windows Update. **K5011257** (FW 4.8) or **KB5011259** (FW 4.7.2).

Microsoft thought it was considered a minor issue. not offering an immediate fix rolled out to everyone.
This KB 5011495 has messed up everywhere. 

Thanks again Mr MS....

Regards
